### DISCUSSION TITLE
Maui Android support for WorkManager

### DISCUSSION THREAD

**Discussion by omxie** [dotnet/maui#4065] [Q&A] [answered]
Hello there, 
is [WorkManager](https://developer.android.com/reference/androidx/work/WorkManager#:~:text=WorkManager%20is%20a%20library%20used,create%20complex%20chains%20of%20work.) available for Maui Android? 
And would this [Xamarin Nuget Package](https://www.nuget.org/packages/Xamarin.Android.Arch.Work.Runtime/) work for Maui? Currently when I install that package, I end up with lots of errors and the application fails to execute.
Any help would be appreciated. 
Thanks!

**Comment by jjonescz** [comment]
Instead of [Xamarin.Android.Arch.Work.Runtime](https://www.nuget.org/packages/Xamarin.Android.Arch.Work.Runtime/), try using [Xamarin.AndroidX.Work.Runtime](https://www.nuget.org/packages/Xamarin.AndroidX.Work.Runtime/). See also http://damien.dennehy.me/using-workmanager-and-listenableworker-in-xamarin-android/.

**Reply by mariglenpupa** [reply]
Any updates on this? I cant get it to work on maui :(
If I dont find a solution soon I'll have to reconsider xamarin

**Comment by BerserkerDotNet** [comment, marked as answer]
I managed to make it work with `<PackageReference Include="Xamarin.AndroidX.Work.Runtime" Version="2.7.1.5" />`
Here is a sample code:
```csharp
    public void SyncData()
    {
        using var builder = new Constraints.Builder();
        builder.SetRequiredNetworkType(NetworkType.Connected);
        var workConstraints = builder.Build();

        var workerRequest = new OneTimeWorkRequest.Builder(typeof(DataSyncWorker))
            .SetConstraints(workConstraints)
            .AddTag(DataSyncWorker.TAG)
            .Build();

        WorkManager.GetInstance(Microsoft.Maui.ApplicationModel.Platform.AppContext)
            .EnqueueUniqueWork(DataSyncWorker.TAG, ExistingWorkPolicy.Keep, workerRequest);
    }
```
and the worker class that implements either `Worker` or `ListenableWorker` as per official docs https://developer.android.com/reference/androidx/work/Worker

**Reply by joseluisct** [reply]
Thank you, it is working now with the version **2.7.1.5** of **Xamarin.AndroidX.Work.Runtime**

**Reply by antmx** [reply]
I got a strange `kotlin.io.path.ExperimentalPathApi` version conflict when I tried the latest version of this (`Xamarin.AndroidX.Work.Runtime` v2.7.1.6).
So I tried referencing `Xamarin.AndroidX.Work.Work.Runtime.Ktx` v2.7.1.6 instead and it all works well now. Thanks for pointing me in the right direction though.

**Comment by ComptonAlvaro** [comment]
I am interesting to run a background task, and it seems that the Worker could be a solution for me. But I am new and I am a bit lose by the moment.

In the asnwered solution, the SyncData() method, where is implemented? In a class insde Platforms\Android folder or in the MAUI folder? I mean if it is an especific class of Android or a generic class.

could it be possible you could share the code that works for you?



Thanks so much.

**Comment by ozersenol** [comment]
is there a full example for this?
